### PR TITLE
EVP_{CIPHER,MD}_CTX_ctrl(): make extra sure to return 0 or 1

### LINKED ISSUE #10106: restore non-negative return value of EVP_CIPHER_CTX_*() in master
Issue is mainly for EVP_CIPHER_CTX_ctrl() but review of history shows that may some other functions are impacted - start is commit 13273237a65d46186b6bea0b51aec90670d4598a , dated  Tue, 9 Jul 2019 07:32:16 +0200.

Before switch to provider most of public function are "boolean like", i.e. never return negative value and zero as error indicator. In consequence many projects use "if ( !foo() )" calls to check for error.

Changes in this are of widely user API is like to open "Pandora's box".
Please restore original "preprovider" functionality.

I also think that in case of error public function should return -1 or negative value but this could be done with new api EVP_XCIPHER...

### PR DESCRIPTION
**Pull request by levitte**
The previous fix (#10108) was incomplete.

Fixes #10106

### CODE REVIEW

**Comment by levitte** [comment]
(note that I haven't really tested.  It's Sunday...)

**Comment by petrovr** [comment]
Richard Levitte wrote:
> The previous fix (#10108) was incomplete.
>
> Fixes #10106

In pre-master EVP_CTRL_INIT returns 1. Just goto will use default value set ate beginning of function (-1).
I think that for this case code should be ret = 1; goto concluide .

Roumen

File: crypto/evp/evp_enc.c [hunk 1]
```diff
@@ -1090,7 +1090,7 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
     case EVP_CTRL_SET_PIPELINE_OUTPUT_BUFS: /* Used by DASYNC */
     case EVP_CTRL_INIT: /* TODO(3.0) Purely legacy, no provider counterpart */
     default:
-        return EVP_CTRL_RET_UNSUPPORTED;
+        goto conclude;
```
Comment (slontis) [inline crypto/evp/evp_enc.c]: Is this supposed to be ret = UNSUPPORTED here??

Comment (t8m) [reply crypto/evp/evp_enc.c]: It is already set to that value.

Comment (levitte) [reply crypto/evp/evp_enc.c]: Yup.  Look at line 1065

Comment (richsalz) [reply crypto/evp/evp_enc.c]: While you're here, please fix the inconsistency and rename the label to end or err like everywhere else in openssl.

Comment (levitte) [reply crypto/evp/evp_enc.c]: Meh :unamused: ...  but ok

**Comment by t8m** [comment]
Is this still WIP?

**Comment by levitte** [comment]
@petrovr, I just added a commit that deals with EVP_CTRL_INIT as you requested.  Please confirm that this is what you thought it should do.

**Comment by levitte** [comment]
And *now* it's time to take this out of WIP :wink:

**Review by slontis — APPROVED** [review]

**Comment by petrovr** [comment]
Richard Levitte wrote:
> @petrovr, I just added a commit that deals with EVP_CTRL_INIT as you requested.  Please confirm that this is what you thought it should do.
> 
Ok.
For protocol currently it is used in crypto/evp/evp_enc.c, line 442, 
legacy code , when flag EVP_CIPH_CTRL_INIT is set and ... etc

Roumen

**Comment by levitte** [comment]
Thanks everyone!

Merged.

6a36f209bcfbc958e30d544c3031ed002dbfe582 EVP_{CIPHER,MD}_CTX_ctrl(): make extra sure to return 0 or 1
d6d74cf4a44f08ed63d942b103198a1e3294295a For provided ciphers, EVP_CIPHER_CTX_ctrl() with EVP_CTRL_INIT always returns 1

**Pull request closed without merging**